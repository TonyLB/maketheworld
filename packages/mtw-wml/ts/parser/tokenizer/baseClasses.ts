import SourceStream from "./sourceStream";

export type TokenBase = {
    source: string;
    startIdx: number;
    endIdx: number;
}

export type TokenComment = {
    type: 'Comment'
} & TokenBase

export type TokenWhitespace = {
    type: 'Whitespace'
} & TokenBase

export type TokenExpressionValue = {
    type: 'ExpressionValue'
} & TokenBase

export type TokenKeyValue = {
    type: 'KeyValue';
    key: string;
} & TokenBase

export type TokenLiteralValue = {
    type: 'LiteralValue';
    value: string;
} & TokenBase

export type TokenProperty = {
    type: 'Property'
} & TokenBase

export type TokenBeginTagOpen = {
    type: 'BeginTagOpen'
} & TokenBase

export type TokenBeginTagClose = {
    type: 'BeginTagClose';
} & TokenBase

export type TokenEndTag = {
    type: 'EndTag';
} & TokenBase

export type TokenDescription = {
    type: 'Description';
} & TokenBase

export type TokenError = {
    type: 'Error';
    message: string;
} & TokenBase

export type Token = TokenComment |
    TokenWhitespace |
    TokenExpressionValue |
    TokenKeyValue |
    TokenLiteralValue |
    TokenProperty |
    TokenBeginTagOpen |
    TokenBeginTagClose |
    TokenEndTag |
    TokenDescription |
    TokenError

export type Tokenizer<T extends TokenBase> = {
    (sourceStream: SourceStream): T | TokenError | undefined
}

export const isTokenError = (item: TokenBase & { type: string }): item is TokenError => (item.type === 'Error')
