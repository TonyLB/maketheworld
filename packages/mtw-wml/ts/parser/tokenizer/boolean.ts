import {
    Tokenizer,
    TokenBooleanProperty
} from "./baseClasses"

export const booleanPropertyTokenizer: Tokenizer<TokenBooleanProperty> = (sourceStream) => {
    const firstChar = sourceStream.lookAhead(1)
    if (firstChar.match(/[A-Za-z\!]/)) {
        const startIdx = sourceStream.position
        let returnValue = firstChar === '!' ? sourceStream.consume(1) : ''
        let key = ''
        while(sourceStream.lookAhead(1).match(/[A-Za-z]/)) {
            key = key + sourceStream.consume(1)
        }
        returnValue = returnValue + key
        return {
            type: 'BooleanValue',
            source: returnValue,
            value: firstChar !== '!',
            key,
            startIdx,
            endIdx: sourceStream.position - 1
        }
    }
    else {
        return undefined
    }
}

export default booleanPropertyTokenizer
