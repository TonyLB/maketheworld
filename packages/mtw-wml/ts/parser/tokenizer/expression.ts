import {
    TokenBase,
    Tokenizer,
    TokenExpressionValue
} from "./baseClasses"
import { expressionStringLiteralSubTokenizer } from "./literal"
import { checkSubTokenizer } from "./utils"

type TokenTemplateString = {
    type: 'TemplateString'
} & TokenBase

export const expressionTemplateStringSubTokenizer: Tokenizer<TokenTemplateString> = (sourceStream) => {
    const firstChar = sourceStream.lookAhead(1)
    if (firstChar === "`") {
        const startIdx = sourceStream.position
        let returnValue = sourceStream.consume(1)
        while(!sourceStream.lookAhead(firstChar)) {
            const nextChar = sourceStream.lookAhead(1)
            if (nextChar === '\\') {
                returnValue = returnValue + sourceStream.consume(2)
            }
            else {
                returnValue = returnValue + sourceStream.consume(1)
                if (nextChar === '$') {
                    const checkExpressionToken = checkSubTokenizer({
                        currentBuffer: returnValue,
                        startIdx,
                        subTokenizer: expressionValueTokenizer,
                        sourceStream
                    })
                    if (checkExpressionToken.error) {
                        return checkExpressionToken.error
                    }
                    else if (checkExpressionToken.returnBuffer) {
                        returnValue = checkExpressionToken.returnBuffer
                    }
                    else {
                        returnValue = returnValue + sourceStream.consume(1)
                    }
                }
            }
            if (sourceStream.isEndOfSource) {
                return {
                    type: 'Error',
                    message: 'Unbounded template string',
                    source: returnValue,
                    startIdx,
                    endIdx: sourceStream.position - 1
                }
            }
        }
        returnValue = returnValue + sourceStream.consume(1)
        return {
            type: 'TemplateString',
            source: returnValue,
            startIdx,
            endIdx: sourceStream.position - 1
        }
    }
    else {
        return undefined
    }
}

export const expressionValueTokenizer: Tokenizer<TokenExpressionValue> = (sourceStream) => {
    if (!sourceStream.lookAhead('{')) {
        return undefined
    }
    let returnValue = ''
    const startIdx = sourceStream.position
    returnValue = returnValue + sourceStream.consume(1)
    while(!sourceStream.lookAhead('}')) {
        const checkLiteral = checkSubTokenizer({
            currentBuffer: returnValue,
            startIdx,
            subTokenizer: expressionStringLiteralSubTokenizer,
            sourceStream
        })
        if (checkLiteral.error) {
            return checkLiteral.error
        }
        if (checkLiteral.returnBuffer) {
            returnValue = checkLiteral.returnBuffer
            continue
        }
        const checkTemplate = checkSubTokenizer({
            currentBuffer: returnValue,
            startIdx,
            subTokenizer: expressionTemplateStringSubTokenizer,
            sourceStream
        })
        if (checkTemplate.error) {
            return checkTemplate.error
        }
        if (checkTemplate.returnBuffer) {
            returnValue = checkTemplate.returnBuffer
            continue
        }
        const checkExpression = checkSubTokenizer({
            currentBuffer: returnValue,
            startIdx,
            subTokenizer: expressionValueTokenizer,
            sourceStream
        })
        if (checkExpression.error) {
            return checkExpression.error
        }
        if (checkExpression.returnBuffer) {
            returnValue = checkExpression.returnBuffer
            continue
        }
        returnValue = returnValue + sourceStream.consume(1)
    }
    returnValue = returnValue + sourceStream.consume(1)
    return {
        type: 'ExpressionValue',
        source: returnValue,
        startIdx,
        endIdx: sourceStream.position - 1
    }
}

export default expressionValueTokenizer
