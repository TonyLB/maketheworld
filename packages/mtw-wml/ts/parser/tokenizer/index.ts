import {
    isTokenError,
    Token,
    TokenDescription,
    TokenTagClose,
    TokenTagOpenBegin,
    TokenTagOpenEnd,
    TokenWhitespace,
    TokenProperty,
    TokenBooleanProperty,
    TokenLiteralValue,
    TokenKeyValue,
    TokenExpressionValue
} from "./baseClasses"
import booleanPropertyTokenizer from "./boolean"
import expressionValueTokenizer from "./expression"
import keyValueTokenizer from "./key"
import literalValueTokenizer from "./literal"
import SourceStream from "./sourceStream"
import tagCloseTokenizer from "./tagClose"

import { tagOpenBeginTokenizer, tagOpenEndTokenizer } from "./tagOpen"
import tagPropertyTokenizer from "./tagProperty"
import whiteSpaceTokenizer from "./whiteSpace"
import { checkSubTokenizers } from "./utils"

enum TokenizerMode {
    Contents,
    Tag
}

export const tokenizer = (sourceStream: SourceStream): Token[] => {
    let mode: TokenizerMode = TokenizerMode.Contents
    let currentDescription: TokenDescription | undefined
    let returnValue: Token[] = []

    while(!sourceStream.isEndOfSource) {
        if (mode === TokenizerMode.Contents) {
            const checkSubTokens = checkSubTokenizers<TokenWhitespace | TokenTagClose | TokenTagOpenBegin>({
                sourceStream,
                subTokenizers: [whiteSpaceTokenizer, tagCloseTokenizer, tagOpenBeginTokenizer],
                callback: (token) => {
                    returnValue.push(token)
                    if (token.type === 'TagOpenBegin') {
                        mode = TokenizerMode.Tag
                    }
                }
            })
            if (checkSubTokens) {
                if (checkSubTokens.success === false) {
                    return [checkSubTokens.error]
                }
                continue
            }
            return [{
                type: 'Error',
                startIdx: sourceStream.position,
                endIdx: sourceStream.position,
                source: sourceStream.lookAhead(1),
                message: 'Unexpected token'
            }]
        }
        else {
            const checkSubTokens = checkSubTokenizers<TokenTagOpenEnd | TokenWhitespace | TokenProperty | TokenBooleanProperty>({
                sourceStream,
                subTokenizers: [tagOpenEndTokenizer, whiteSpaceTokenizer, tagPropertyTokenizer, booleanPropertyTokenizer],
                callback: (token) => {
                    returnValue.push(token)
                    if (token.type === 'TagOpenEnd') {
                        mode = TokenizerMode.Contents
                    }
                    if (token.type === 'Property') {
                        //
                        // Consume any whitespace first...
                        //
                        const checkWhitespace = whiteSpaceTokenizer(sourceStream)
                        if (checkWhitespace) {
                            if (isTokenError(checkWhitespace)) {
                                return {
                                    success: false,
                                    error: checkWhitespace
                                }
                            }
                            returnValue.push(checkWhitespace)
                        }
                        //
                        // ... and then confirm that a value follows the key
                        //
                        const checkValue = checkSubTokenizers<TokenLiteralValue | TokenKeyValue | TokenExpressionValue>({
                            sourceStream,
                            subTokenizers: [literalValueTokenizer, keyValueTokenizer, expressionValueTokenizer],
                            callback: (token) => {
                                returnValue.push(token)
                            }
                        })
                        if (checkValue) {
                            if (checkValue.success === false) {
                                return checkValue
                            }
                        }
                        else {
                            return {
                                success: false,
                                error: {
                                    type: 'Error',
                                    source: sourceStream.lookAhead(1),
                                    startIdx: sourceStream.position,
                                    endIdx: sourceStream.position,
                                    message: 'Unexpected token'
                                }
                            }
                        }
                    }
                }
            })
            if (checkSubTokens) {
                if (checkSubTokens.success === false) {
                    return [checkSubTokens.error]
                }
                continue
            }
            return [{
                type: 'Error',
                source: sourceStream.lookAhead(1),
                startIdx: sourceStream.position,
                endIdx: sourceStream.position,
                message: 'Unexpected token'
            }]
        }
    }
    return returnValue
}

export default tokenizer
