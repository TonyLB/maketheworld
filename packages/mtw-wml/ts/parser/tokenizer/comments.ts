import { Tokenizer, TokenComment } from "./baseClasses"

export const commentTokenizer: Tokenizer<TokenComment> = (sourceStream) => {
    if (sourceStream.lookAhead('//')) {
        const nextInstance = sourceStream.nextInstance(['\n'])
        const startIdx = sourceStream.position
        if (nextInstance > 1) {
            sourceStream.consume(nextInstance)
        }
        else {
            sourceStream.consume(sourceStream.source.length - sourceStream.position)
        }
        const endIdx = sourceStream.position - 1
        return {
            type: 'Comment',
            startIdx,
            endIdx
        }
    }
    if (sourceStream.lookAhead('/*')) {
        const nextInstance = sourceStream.nextInstance(['*/'])
        const startIdx = sourceStream.position
        if (nextInstance > 1) {
            sourceStream.consume(nextInstance + 2)
        }
        else {
            sourceStream.consume(sourceStream.source.length - sourceStream.position)
        }
        const endIdx = sourceStream.position - 1
        if (nextInstance === -1) {
            return {
                type: 'Error',
                message: 'Unbounded comment',
                startIdx,
                endIdx
            }
        }
        else {
            return {
                type: 'Comment',
                startIdx,
                endIdx
            }
        }
    }
    return undefined
}

export default commentTokenizer
