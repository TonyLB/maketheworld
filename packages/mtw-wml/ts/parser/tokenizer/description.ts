import { Tokenizer, TokenDescription } from "./baseClasses"

export const descriptionTokenizer: Tokenizer<TokenDescription> = (sourceStream) => {
    const startIdx = sourceStream.position
    let value = ''
    while (![' ', '\n', '\t', '\r', '<'].includes(sourceStream.lookAhead(1))) {
        if (['/*', '//'].includes(sourceStream.lookAhead(2))) {
            break
        }
        if (sourceStream.isEndOfSource) {
            break
        }
        if (sourceStream.lookAhead(1) === '\\') {
            sourceStream.consume(1)
        }
        value = `${value}${sourceStream.consume(1)}`
    }
    const endIdx = sourceStream.position
    if (startIdx === endIdx) {
        return undefined
    }
    else {
        return {
            type: 'Description',
            startIdx,
            endIdx,
            value
        }
    }
}

export default descriptionTokenizer
