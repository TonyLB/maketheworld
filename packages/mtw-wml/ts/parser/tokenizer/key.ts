import { Tokenizer, TokenKeyValue } from "./baseClasses"
import whiteSpaceTokenizer from "./whiteSpace"
import { checkSubTokenizers } from "./utils"

export const keyValueTokenizer: Tokenizer<TokenKeyValue> = (sourceStream) => {
    const firstChar = sourceStream.lookAhead(1)
    if (firstChar === '(') {
        const startIdx = sourceStream.position
        sourceStream.consume(1)
        const checkWhitespaceOne = checkSubTokenizers({
            sourceStream,
            subTokenizers: [whiteSpaceTokenizer],
            callback: (token) => {}
        })
        if (checkWhitespaceOne && (checkWhitespaceOne.success === false)) {
            return checkWhitespaceOne.error
        }
        const valueStartIdx = sourceStream.position
        if (sourceStream.lookAhead(1).match(/[A-Za-z\_]/)) {
            sourceStream.consume(1)
            while(sourceStream.lookAhead(1).match(/[A-Za-z0-9\_]/)) {
                sourceStream.consume(1)
            }    
        }
        const value = sourceStream.source.slice(valueStartIdx, sourceStream.position)
        const checkWhitespaceTwo = checkSubTokenizers({
            sourceStream,
            subTokenizers: [whiteSpaceTokenizer],
            callback: (token) => {}
        })
        if (checkWhitespaceTwo && (checkWhitespaceTwo.success === false)) {
            return checkWhitespaceTwo.error
        }
        if (!sourceStream.lookAhead(')')) {
            return {
                type: 'Error',
                source: sourceStream.lookAhead(1),
                startIdx: sourceStream.position,
                endIdx: sourceStream.position,
                message: 'Unexpected token'
            }
        }
        sourceStream.consume(1)
        if (sourceStream.position !== startIdx) {
            return {
                type: 'KeyValue',
                value,
                startIdx,
                endIdx: sourceStream.position - 1
            }
        }
    }
    return undefined
}

export default keyValueTokenizer
