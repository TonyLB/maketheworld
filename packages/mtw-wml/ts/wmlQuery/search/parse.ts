import { ParseException } from '../../parser/baseClasses'
import { SearchParse, SearchToken, SearchParseEvaluation, SearchParseComma, SearchParseExplicitGroupOpen } from './baseClasses'

function findLastIndex<T>(array: Array<T>, predicate: (value: T, index: number, obj: T[]) => boolean): number {
    let l = array.length;
    while (l--) {
        if (predicate(array[l], l, array))
            return l;
    }
    return -1;
}

const reduceStack = (stack: (SearchParse | SearchParseComma)[], index: number): SearchParse[] => {
    if (stack.find(({ type }) => (type === 'Comma'))) {
        let buffer: SearchParse[] = []
        let returnValues: SearchParse[] = []
        stack.forEach((item) => {
            if (isSearchComma(item)) {
                if (buffer.length) {
                    if (buffer.length === 1) {
                        returnValues.push(buffer[0])
                    }
                    else {
                        returnValues.push({
                            type: 'Serial',
                            items: buffer
                        })
                    }
                }
                buffer = []
            }
            else {
                buffer.push(item)
            }
        })
        const lastItems: SearchParse[] = buffer.length
            ? buffer.length === 1 ? buffer : [{
                    type: 'Serial',
                    items: buffer
                }]
            : []
        return [{
            type: 'Parallel',
            items: [
                ...returnValues,
                ...lastItems
            ]
        }]
    }
    else {
        return stack.filter(isSearchParse)
    }
}

const isSearchParse = (value: SearchParseEvaluation): value is SearchParse => (['Tag', 'Property', 'First', 'NthChild', 'Serial', 'Parallel'].includes(value.type))
const isSearchComma = (value: SearchParseEvaluation): value is SearchParseComma => (value.type === 'Comma')
const isSearchParseOrComma = (value: SearchParseEvaluation): value is (SearchParse | SearchParseComma) => (value.type === 'Comma' || isSearchParse(value))
const isSearchParseExplicitOpen = (value: SearchParseEvaluation): value is SearchParseExplicitGroupOpen => (value.type === 'ExplicitOpen')

export const searchParse = (tokens: SearchToken[]): SearchParse[] => {
    let stack: SearchParseEvaluation[] = []
    tokens.forEach((token, index) => {
        switch(token.type) {
            case 'Tag':
                stack.push({
                    type: 'Tag',
                    tag: token.tag
                })
                break
            case 'Property':
                stack.push({
                    type: 'Property',
                    key: token.key,
                    value: token.value
                })
                break
            case 'First':
                stack.push({
                    type: 'First'
                })
                break
            case 'NthChild':
                stack.push({
                    type: 'NthChild',
                    n: token.n
                })
                break
            case 'GroupOpen':
                stack.push({
                    type: 'ExplicitOpen',
                    token: index
                })
                break
            case 'Comma':
                stack.push({
                    type: 'Comma'
                })
                break
            case 'GroupClose':
                const groupOpenIndex = findLastIndex(stack, ({ type }) => (type === 'ExplicitOpen'))
                if (groupOpenIndex === -1) {
                    throw new ParseException('Unexpected group closure', index, index)
                }
                else {
                    const groupItems = stack.slice(groupOpenIndex + 1) || []
                    stack = [
                        ...(groupOpenIndex > 0 ? stack.slice(0, groupOpenIndex) : []),
                        ...reduceStack(groupItems.filter(isSearchParseOrComma), index)
                    ]
                }
                break
            default:
                throw new ParseException(`Unexpected token: "${token.type}"`, index, index)
        }
    })
    const groupOpen = stack.find(isSearchParseExplicitOpen)
    if (groupOpen) {
        throw new ParseException('Unmatched group open', groupOpen.token, groupOpen.token)
    }
    return reduceStack(stack.filter(isSearchParseOrComma), stack.length) || []
}

export default searchParse
