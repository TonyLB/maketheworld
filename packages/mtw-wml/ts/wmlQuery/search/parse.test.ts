import parse from './parse'
import SourceStream from '../../parser/tokenizer/sourceStream'
import tokenizer from './tokenize'

describe('wmlQuery select parse', () => {
    beforeEach(() => {
        jest.clearAllMocks()
        jest.resetAllMocks()
    })

    it('should return empty list on empty source', () => {
        const testTokens = []
        expect(parse(testTokens)).toEqual([])
    })
    it('should parse a single sequence', () => {
        const testTokens = tokenizer(new SourceStream('Asset Room:first Exit'))
        expect(parse(testTokens)).toMatchSnapshot()
    })
    it('should parse a singly nested sub-sequence', () => {
        const testTokens = tokenizer(new SourceStream('(Asset, Story) Room:first Exit'))
        expect(parse(testTokens)).toMatchSnapshot()
    })
    //
    // TODO: Fix buggy results on deeply nested sub-sequence
    //
    it('should parse a deeply nested sub-sequence', () => {
        const testTokens = tokenizer(new SourceStream('(Asset, Story) Room:first Exit, Asset (Room Exit, (Room, Feature) Description)'))
        expect(parse(testTokens)).toMatchSnapshot()
    })

})
