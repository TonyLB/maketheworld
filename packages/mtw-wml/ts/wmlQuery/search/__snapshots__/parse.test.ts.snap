// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`wmlQuery select parse should parse a deeply nested sub-sequence 1`] = `
Array [
  Object {
    "items": Array [
      Object {
        "items": Array [
          Object {
            "items": Array [
              Object {
                "tag": "Asset",
                "type": "Tag",
              },
              Object {
                "tag": "Story",
                "type": "Tag",
              },
            ],
            "type": "Parallel",
          },
          Object {
            "tag": "Room",
            "type": "Tag",
          },
          Object {
            "type": "First",
          },
          Object {
            "tag": "Exit",
            "type": "Tag",
          },
        ],
        "type": "Serial",
      },
      Object {
        "items": Array [
          Object {
            "tag": "Asset",
            "type": "Tag",
          },
          Object {
            "items": Array [
              Object {
                "items": Array [
                  Object {
                    "tag": "Room",
                    "type": "Tag",
                  },
                  Object {
                    "tag": "Exit",
                    "type": "Tag",
                  },
                ],
                "type": "Serial",
              },
              Object {
                "items": Array [
                  Object {
                    "items": Array [
                      Object {
                        "tag": "Room",
                        "type": "Tag",
                      },
                      Object {
                        "tag": "Feature",
                        "type": "Tag",
                      },
                    ],
                    "type": "Parallel",
                  },
                  Object {
                    "tag": "Description",
                    "type": "Tag",
                  },
                ],
                "type": "Serial",
              },
            ],
            "type": "Parallel",
          },
        ],
        "type": "Serial",
      },
    ],
    "type": "Parallel",
  },
]
`;

exports[`wmlQuery select parse should parse a single sequence 1`] = `
Array [
  Object {
    "tag": "Asset",
    "type": "Tag",
  },
  Object {
    "tag": "Room",
    "type": "Tag",
  },
  Object {
    "type": "First",
  },
  Object {
    "tag": "Exit",
    "type": "Tag",
  },
]
`;

exports[`wmlQuery select parse should parse a singly nested sub-sequence 1`] = `
Array [
  Object {
    "items": Array [
      Object {
        "tag": "Asset",
        "type": "Tag",
      },
      Object {
        "tag": "Story",
        "type": "Tag",
      },
    ],
    "type": "Parallel",
  },
  Object {
    "tag": "Room",
    "type": "Tag",
  },
  Object {
    "type": "First",
  },
  Object {
    "tag": "Exit",
    "type": "Tag",
  },
]
`;
